lumicks/pylake/population/mixture.py

```diff
@@ -0,0 +1,187 @@
+import numpy as np
+from sklearn.mixture import GaussianMixture
+from scipy import stats
+
+
+def as_sorted(fcn):
+    """Decorator to return results sorted according to mapping array."""
+
+    def wrapper(self, *args, **kwargs):
+        result = fcn(self, *args, **kwargs)
+        return result[self._map]
+
+    wrapper.__doc__ = fcn.__doc__
+
+    return wrapper
+
+
+class GaussianMixtureModel:
+    """A wrapper around scikit-learn's GMM. This model accepts a 1D
+    array as training data. The state parameters are sorted according
+    to state mean in order to facilitate comparison of models between
+    different datasets.
+
+    Parameters
+    ----------
+    data : array-like
+        data object used for model training
+    n_states : int
+        number of Gaussian components in the model
+    init_method : 'kmeans' or 'random'
+        method used to initialize parameters
+    n_init : int
+        number of initializations to perform
+    tol : float
+        convergence tolerance
+    max_iter : int
+        maximum number of iterations to perform
+    """
+
+    def __init__(self, data, n_states, init_method, n_init, tol, max_iter):
+        self.n_states = n_states
+        self._model = GaussianMixture(
+            n_components=n_states,
+            init_params=init_method,
+            n_init=n_init,
+            tol=tol,
+            max_iter=max_iter,
+        )
+        self._model.fit(np.reshape(data, (-1, 1)))
+        self._bic = self._model.bic(np.reshape(data, (-1, 1)))
+        self._aic = self._model.aic(np.reshape(data, (-1, 1)))
```

Comment:
* To avoid needless repetition, you should pull this out into a variable. That way it's easy to see that all 3 of these are getting the same data. Otherwise, readers need to spend more time to confirm that all 3 are indeed identical.
* Are there any performance issues evaluating `bic` and `aic` eagerly here?

Reply:
Perhaps for extremely large datasets. The alternative though is to keep a reference to the training data, which in the current form I'm not sure is particularly useful either.

We could go back to how I had originally had the API set up where `aic` and `bic` are methods and you would supply the data as an argument. This is how it is implemented in `sklearn` and is more flexible, but as Joep pointed out it doesn't necessarily make sense to calculate the information criteria for data that wasn't used to train the model.

Reply:
I'm skeptical whether `aic` or `bic` for out of sample testing makes sense. This is certainly not how it was used in the papers that introduced these methods. I'd like to see a reference where this is motivated because I see potential for user error if they start supplying it with different data than what they trained on.

As for perf impact, it should only be a matter of evaluating the likelihood once for each (if it is not actually cached by sklearn internally), so I wouldn't expect a big performance overhead.

Reply:
OK, if it ends up being a performance issue, it easy enough to make it a property.